File: python_workflow/tests/test_io_utils.py
import pandas as pd
import pytest

from lexsync.io_utils import (
    _is_continuous,
    clean_field,
    read_config,
    read_csv_utf8,
    sha256_file,
    slugify,
    write_csv_utf8,
)

# The R engine (readr) writes LF on every platform, so LF is the cross-engine
# contract. test-io_utils.R pins these same bytes and this same digest.
EXPECTED_BYTES = b"word,condition\ncat,high\ndog,low\n"
EXPECTED_SHA256 = "8084f4888fa65455ee56e7eca2954b07b114f5b962bb78c6e8c73c9928e66ad5"


def _frame():
    return pd.DataFrame({"word": ["cat", "dog"], "condition": ["high", "low"]})


def test_write_csv_utf8_pins_lf_on_every_platform(tmp_path):
    path = write_csv_utf8(_frame(), str(tmp_path / "s.csv"))
    assert open(path, "rb").read() == EXPECTED_BYTES


def test_write_csv_utf8_digest_is_platform_independent(tmp_path):
    # The datasheet advertises this digest as provenance, so it must be fixed by
    # the content alone -- not by the os.linesep of the machine that ran it.
    path = write_csv_utf8(_frame(), str(tmp_path / "s.csv"))
    assert sha256_file(path) == EXPECTED_SHA256


def test_write_csv_utf8_round_trips(tmp_path):
    path = write_csv_utf8(_frame(), str(tmp_path / "nested" / "s.csv"))
    assert read_csv_utf8(path).equals(_frame())


def test_read_csv_utf8_keeps_words_that_look_like_missing_values(tmp_path):
    path = tmp_path / "s.csv"
    path.write_bytes(b"word,gloss\nnull,x\nnan,y\n,z\nNA,w\n")
    out = read_csv_utf8(str(path))
    assert list(out["word"][:2]) == ["null", "nan"]
    assert out["word"][2:].isna().all()


def test_clean_field_rejects_control_characters():
    with pytest.raises(ValueError, match="control characters"):
        clean_field("cat\ndog", "word")


def test_clean_field_allows_commas_and_quotes():
    assert clean_field('the "big" cat, asleep', "word") == 'the "big" cat, asleep'


def test_slugify_is_lowercase_and_path_safe():
    assert slugify("En Lexdec", "English!") == "en_lexdec_english"


# Pins the same contract as "slugify folds case whatever the locale" in the R
# engine's test-io_utils.R, where base tolower() folded "I" to the dotless "i"
# under a Turkish locale and wrote the design's artifacts under a name this
# engine never produces. str.lower() is already locale-invariant; the assertion
# is what the R engine is now held to.
def test_slugify_is_locale_invariant():
    assert slugify("STUDY_I") == "study_i"
    assert slugify("En Lexdec", "English!").isascii()


def test_read_config_refuses_a_duplicated_mapping_key(tmp_path):
    # yaml.safe_load keeps the last value for a repeated key, silently; the R
    # engine's yaml::read_yaml() rejects the key by its libyaml parser default,
    # so this engine accepted a config the R engine refused. The messages differ
    # because R's comes from the C parser; behavioural parity -- both engines
    # refuse -- is the contract, and test-io_utils.R pins the R side of it.
    path = tmp_path / "dup.yaml"
    path.write_text("name: a\nname: b\n", encoding="utf-8")
    with pytest.raises(ValueError, match="duplicate mapping key 'name'"):
        read_config(str(path))
    # A key repeated in a NESTED mapping must be caught too, not just at the top.
    path.write_text("items:\n  source: table\n  source: pool\n", encoding="utf-8")
    with pytest.raises(ValueError, match="duplicate mapping key 'source'"):
        read_config(str(path))
    # And an ordinary config still loads, so the strict loader costs nothing.
    path.write_text("name: a\nitems:\n  source: pool\n", encoding="utf-8")
    assert read_config(str(path)) == {"name": "a", "items": {"source": "pool"}}


def test_a_continuous_design_over_a_supplied_pool_takes_the_continuous_path():
    # The predicate allowed corpus and table only, so a 'continuous' block over
    # items.source 'pool' fell through to the conditions matcher and failed with
    # a different obscure error in each engine, even though run_pipeline's
    # corpus/pool branch handles continuous selection generically. The R twin
    # "a continuous design over a supplied pool takes the continuous path" in
    # test-io_utils.R pins the same outcomes.
    cont = {"predictor": "frequency", "controls": ["length"]}
    assert _is_continuous({"continuous": cont, "items": {"source": "pool"}})
    assert _is_continuous({"continuous": cont})            # default source: corpus
    assert not _is_continuous({"items": {"source": "pool"}})
    with pytest.raises(ValueError,
                       match="cannot be combined with items.source 'generate'"):
        _is_continuous({"continuous": cont, "items": {"source": "generate"}})

File: R_workflow/tests/testthat/test-io_utils.R
# The Python engine pins LF to match readr's always-LF output; test_io_utils.py
# pins these same bytes and this same digest.
expected_bytes <- charToRaw("word,condition\ncat,high\ndog,low\n")
expected_sha256 <- "8084f4888fa65455ee56e7eca2954b07b114f5b962bb78c6e8c73c9928e66ad5"

make_frame <- function() {
  data.frame(
    word = c("cat", "dog"), condition = c("high", "low"),
    stringsAsFactors = FALSE
  )
}

test_that("write_csv_utf8 pins LF on every platform", {
  path <- file.path(tempfile("lexsync"), "s.csv")
  write_csv_utf8(make_frame(), path)
  expect_identical(readBin(path, "raw", file.size(path)), expected_bytes)
})

test_that("write_csv_utf8 digest is platform-independent", {
  # The datasheet advertises this digest as provenance, so it must be fixed by
  # the content alone -- not by the line terminator of the host platform.
  path <- file.path(tempfile("lexsync"), "s.csv")
  write_csv_utf8(make_frame(), path)
  expect_identical(sha256_file(path), expected_sha256)
})

test_that("write_csv_utf8 round-trips through read_csv_utf8", {
  path <- file.path(tempfile("lexsync"), "nested", "s.csv")
  write_csv_utf8(make_frame(), path)
  expect_equal(as.data.frame(read_csv_utf8(path)), make_frame())
})

# Every artefact-writing path must pin LF, not just the CSV one. write_datasheet and
# write_run_log used bare writeLines(), whose connection is opened in TEXT mode, so on
# Windows all three of their outputs came out CRLF while the Python engine's twins were
# LF: measured 109 CRLF against 0 in one datasheet. The datasheet is the provenance
# artefact of the whole package, so its bytes are the last that should depend on which
# machine happened to build it. The .jsonl run log was already binary and is included
# here so the whole set is covered by one assertion.
test_that("the datasheet and run log pin LF like every other artefact", {
  dir <- tempfile("lf"); dir.create(dir)
  # Fixtures are local: testthat runs each file in its own environment, so the ones in
  # test-datasheet.R are not in scope here.
  schema <- yaml::read_yaml(system.file("extdata", "schema.yaml", package = "lexsync"))
  stim <- data.frame(word = c("cat", "dog"), condition = c("hi", "lo"), set = c(1, 1),
                     length = 3L, frequency = c(6, 3), stringsAsFactors = FALSE)
  design <- list(name = "lf", language = "english", match_on = list("length"),
                 n_per_condition = 1L, counterbalance = list(lists = 1))
  ds <- build_datasheet(design, schema, NULL, stim, "x.csv",
                        list(stimuli = NULL, experiments = list()), 2026, engine = "R")
  paths <- write_datasheet(ds, file.path(dir, "d.json"), file.path(dir, "d.md"))

  log <- new_run_log("lf", meta = list(design = "lf"))
  log <- log_step(log, "a step", list(n = 1L))
  write_run_log(log, file.path(dir, "log.md"), file.path(dir, "log.jsonl"))

  for (p in c(paths, file.path(dir, c("log.md", "log.jsonl")))) {
    raw <- readBin(p, "raw", file.size(p))
    expect_false(as.raw(13) %in% raw, info = basename(p))
    expect_true(as.raw(10) %in% raw, info = basename(p))
  }
})

test_that("read_config refuses a duplicated mapping key", {
  # yaml::read_yaml() rejects duplicate map keys by its libyaml parser default;
  # asserting the refusal here pins that default, so a future yaml-package
  # regression is caught rather than silently reopening the gap. The Python
  # engine's yaml.safe_load kept the last value, so read_config there now refuses
  # too (test_read_config_refuses_a_duplicated_mapping_key in test_io_utils.py).
  # The messages differ because this one comes from the C parser; behavioural
  # parity -- both engines refuse -- is the contract.
  path <- file.path(tempdir(), "dup.yaml")
  writeLines(c("name: a", "name: b"), path)
  expect_error(read_config(path), "Duplicate map key", fixed = TRUE)
  # A key repeated in a NESTED mapping must be caught too, not just at the top.
  writeLines(c("items:", "  source: table", "  source: pool"), path)
  expect_error(read_config(path), "Duplicate map key", fixed = TRUE)
  # And an ordinary config still loads.
  writeLines(c("name: a", "items:", "  source: pool"), path)
  expect_identical(read_config(path), list(name = "a", items = list(source = "pool")))
})

test_that("a continuous design over a supplied pool takes the continuous path", {
  # The predicate allowed corpus and table only, so a 'continuous' block over
  # items.source 'pool' fell through to the conditions matcher and failed with a
  # different obscure error in each engine, even though run_pipeline's corpus/pool
  # branch handles continuous selection generically. The Python twin
  # test_a_continuous_design_over_a_supplied_pool_takes_the_continuous_path in
  # test_io_utils.py pins the same outcomes.
  cont <- list(predictor = "frequency", controls = list("length"))
  expect_true(lexsync:::.is_continuous(list(continuous = cont,
                                            items = list(source = "pool"))))
  expect_true(lexsync:::.is_continuous(list(continuous = cont)))  # default: corpus
  expect_false(lexsync:::.is_continuous(list(items = list(source = "pool"))))
  expect_error(lexsync:::.is_continuous(list(continuous = cont,
                                             items = list(source = "generate"))),
               "cannot be combined with items.source 'generate'", fixed = TRUE)
})

test_that("clean_field rejects control characters but allows commas and quotes", {
  expect_error(clean_field("cat\ndog", "word"), "control characters")
  expect_identical(clean_field('the "big" cat, asleep', "word"), 'the "big" cat, asleep')
})

test_that("slugify is lower-case and path-safe", {
  expect_identical(slugify("En Lexdec", "English!"), "en_lexdec_english")
})

# Pins the same contract as test_slugify_is_locale_invariant in the Python
# engine's test_io_utils.py. Every artefact path in both engines is built from
# slugify(), so a locale-sensitive fold would write this design's files under a
# name the Python engine never produces. Base `tolower()` maps "I" to the
# dotless "i" under a Turkish or Azeri locale, taking the slug out of ASCII --
# and it does so after the reduction to [A-Za-z0-9_], which is why that
# reduction does not make the fold safe by itself.
test_that("slugify folds case whatever the locale", {
  expect_identical(slugify("STUDY_I"), "study_i")

  old <- Sys.getlocale("LC_CTYPE")
  # "Turkish" is the Windows spelling of this locale and resolves nowhere else, so
  # on a Linux runner the test skipped every time and guarded nothing -- which the
  # CI step's zero-skip assertion then reported as a failure. Try each platform's
  # name for the same locale; CI generates tr_TR.UTF-8, so the skip does not fire
  # there and the fold is actually exercised.
  turkish <- ""
  for (nm in c("Turkish", "tr_TR.UTF-8", "tr_TR.utf8", "tr_TR")) {
    turkish <- suppressWarnings(Sys.setlocale("LC_CTYPE", nm))
    if (nzchar(turkish)) break
  }
  skip_if(!nzchar(turkish), "cannot switch to a Turkish locale on this platform")
  on.exit(Sys.setlocale("LC_CTYPE", old), add = TRUE)
  expect_identical(slugify("STUDY_I"), "study_i")
  expect_identical(slugify("En Lexdec", "English!"), "en_lexdec_english")
})
